app/controllers/criteria_controller.rb

```diff
@@ -114,10 +114,21 @@ def update
   # Handles the drag/drop criteria sorting.
   def update_positions
     @assignment = Assignment.find(params[:assignment_id])
+    invalid_criteria = false
 
     ApplicationRecord.transaction do
-      params[:criterion].each_with_index do |id, index|
-        Criterion.update(id, position: index + 1) unless id.blank?
+      Criterion.transaction(requires_new: true) do
+        params[:criterion].each_with_index do |id, index|
+          Criterion.update(id, position: index + 1) unless id.blank?
+          # check if any criteria invalid
+          if Criterion.find(id).assessment_id != @assignment.assessment_id
```

Comment:
This part does work but it's a bit redundant. Instead it's simpler to modify the query on the line above to use an association query (something like `@assignment.criteria.find(id)...`)

This will raise an error if the criterion isn't found for that assignment, and then you can catch that error (using a `rescue`) and handle it using the same code you have below.